module/platform/Webots/src/Webots.cpp

```diff
@@ -292,172 +301,201 @@ namespace module::platform {
 
-        // Now that we are connected, we can set up our reaction handles with this file descriptor and send the sensor
-        // timesteps message
+            fd = tcpip_connect();
 
-        // Create the sensor timesteps message
-        // This will activate all of the sensors in the simulator
-        const std::vector<char> data =
-            NUClear::util::serialise::Serialise<ActuatorRequests>::serialise(create_sensor_time_steps(time_step));
-
-        const uint32_t Nn = htonl(data.size());
-
-        // Send the sensor timesteps message
-        if (send(fd, &Nn, sizeof(Nn), 0) != sizeof(Nn)) {
-            log<NUClear::ERROR>(fmt::format("Error in sending ActuatorRequests' message size,  {}", strerror(errno)));
-            return;
-        }
-        if (send(fd, data.data(), data.size(), 0) != (signed) data.size()) {
-            log<NUClear::ERROR>(fmt::format("Error sending ActuatorRequests message, {}", strerror(errno)));
-            return;
-        }
-
-        // Receiving
-        read_io = on<IO>(fd, IO::READ).then([this]() {
-            // ************************** Receiving ***************************************
-            // Get the size of the message
-            uint32_t Nn;
-            if (recv(fd, &Nn, sizeof(Nn), MSG_WAITALL) != sizeof(Nn)) {
-                log<NUClear::ERROR>("Failed to read message size from TCP connection");
-                return;
-            }
-
-            // Convert from network endian to host endian
-            const uint32_t Nh = ntohl(Nn);
-
-            // Get the message
-            std::vector<char> data(Nh, 0);
-            uint64_t message = uint64_t(recv(fd, data.data(), Nh, MSG_WAITALL));
-            if (message != Nh) {
-                log<NUClear::ERROR>("Failed to read message from TCP connection");
+            if (fd == -1) {
+                // Connection failed
+                log<NUClear::ERROR>("Failed to connect to server.");
+                active_reconnect.store(false);
                 return;
             }
 
-            log<NUClear::TRACE>("Received sensor measurements");
-
-            // If we have data to deserialise then deserialise the message into a neutron
-            if (Nh > 0) {
-                translate_and_emit_sensor(NUClear::util::serialise::Serialise<SensorMeasurements>::deserialise(data));
-            }
-            // If we don't have any data then this should correspond to a message with all fields set to defaults
-            else {
-                translate_and_emit_sensor(SensorMeasurements(SensorMeasurements::protobuf_type()));
-            }
+            // Receiving
+            read_io = on<IO>(fd, IO::READ | IO::CLOSE | IO::ERROR).then("Read Stream", [this](const IO::Event& event) {
+                if ((event.events & IO::READ) != 0) {
+                    // Service the watchdog
+                    emit<Scope::WATCHDOG>(ServiceWatchdog<Webots>());
```

Comment:
I would recommend you service the watchdog only once you process a packet, so that if you start getting garbage you reset